src/darwin/Framework/CHIP/MTRDemuxingStorage.mm

```diff
@@ -0,0 +1,413 @@
+        return DeleteGlobalValue(keyString);
+    }
+
+    if (IsIndexSpecificKey(keyString)) {
+        FabricIndex index;
+        ReturnErrorOnFailure(ExtractIndexFromKey(keyString, &index));
+        ReturnErrorOnFailure(ExtractIndexSpecificKey(keyString, &keyString));
+        return DeleteIndexSpecificValue(index, keyString);
+    }
+
+    return CHIP_ERROR_INVALID_ARGUMENT;
+}
+
+NSData * _Nullable MTRDemuxingStorage::GetGlobalValue(NSString * key)
+{
+    if (IsMemoryOnlyGlobalKey(key)) {
+        return GetInMemoryValue(key);
+    }
+
+    MTR_LOG_ERROR("MTRDemuxingStorage reading unknown global key: %@", key);
+
+    return nil;
+}
+
+NSData * _Nullable MTRDemuxingStorage::GetIndexSpecificValue(FabricIndex index, NSString * key)
+{
+    if (IsMemoryOnlyIndexSpecificKey(key)) {
+        return GetInMemoryValue(FullyQualifiedKey(index, key));
+    }
+
+    return nil;
+}
+
+CHIP_ERROR MTRDemuxingStorage::SetGlobalValue(NSString * key, NSData * data)
+{
+    if (IsMemoryOnlyGlobalKey(key)) {
+        // Fabric index list only needs to be stored in-memory, not persisted,
+        // because we do not tie controllers to specific fabric indices.
+        return SetInMemoryValue(key, data);
+    }
+
+    MTR_LOG_ERROR("MTRDemuxingStorage setting unknown global key: %@", key);
+
+    return CHIP_ERROR_PERSISTED_STORAGE_FAILED;
+}
+
+CHIP_ERROR MTRDemuxingStorage::SetIndexSpecificValue(FabricIndex index, NSString * key, NSData * data)
+{
+    if ([key isEqualToString:@"n"]) {
+        auto * controller = [mFactory runningControllerForFabricIndex:index];
+        if (controller == nil) {
+            return CHIP_ERROR_PERSISTED_STORAGE_FAILED;
+        }
+
+        ReturnErrorOnFailure([controller.controllerDataStore storeLastUsedNOC:data]);
+    }
+
+    if (IsMemoryOnlyIndexSpecificKey(key)) {
+        return SetInMemoryValue(FullyQualifiedKey(index, key), data);
+    }
+
+    return CHIP_ERROR_PERSISTED_STORAGE_FAILED;
+}
+
+CHIP_ERROR MTRDemuxingStorage::DeleteGlobalValue(NSString * key)
+{
+    if (IsMemoryOnlyGlobalKey(key)) {
+        return DeleteInMemoryValue(key);
+    }
+
+    MTR_LOG_ERROR("MTRDemuxingStorage deleting unknown global key: %@", key);
+
+    return CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND;
+}
+
+CHIP_ERROR MTRDemuxingStorage::DeleteIndexSpecificValue(FabricIndex index, NSString * key)
+{
+    if (IsMemoryOnlyIndexSpecificKey(key)) {
+        return DeleteInMemoryValue(FullyQualifiedKey(index, key));
+    }
+
+    return CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND;
+}
+
+NSData * _Nullable MTRDemuxingStorage::GetInMemoryValue(NSString * key) { return mInMemoryStore[key]; }
+
+CHIP_ERROR MTRDemuxingStorage::SetInMemoryValue(NSString * key, NSData * data)
+{
+    mInMemoryStore[key] = data;
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR MTRDemuxingStorage::DeleteInMemoryValue(NSString * key)
+{
+    BOOL present = (mInMemoryStore[key] != nil);
+    [mInMemoryStore removeObjectForKey:key];
```

Comment:
This would be a no op if the key is not present but would it be good to-  if (present) { [mInMemoryStore removeObjectForKey:key]; }

Reply:
Done.